README.md
[![Travis-CI Build Status](https://travis-ci.org/julian-urbano/ircor.svg?branch=develop)](https://travis-ci.org/julian-urbano/ircor)
[![License](https://img.shields.io/badge/license-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)
[![CRAN version](http://www.r-pkg.org/badges/version/ircor?color=blue)](https://cran.r-project.org/package=ircor) 
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/grand-total/ircor?color=blue)](https://cran.r-project.org/package=ircor) 

# ircor

Provides implementation of various correlation coefficients of common use in Information Retrieval, such as Kendall and AP correlation coefficients, with and without ties.

For reference please refer to *Julián Urbano and Mónica Marrero, "[The Treatment of Ties in AP Correlation](http://julian-urbano.info/files/publications/072-treatment-ties-ap-correlation.pdf)", ACM ICTIR, 2017*.

## Installation

You may install the stable release from CRAN

```r
install.packages("ircor")
```

or the latest development version from GitHub

```r
devtools::install_github("julian-urbano/ircor", ref = "develop")
```

## Usage

`tau` and `tauAP` implement the Kendall tau and Yilmaz tauAP correlation coefficients, where no ties are allowed between items:

```r
x <- c(0.06, 0.2, 0.27, 0.37, 0.57, 0.63, 0.66, 0.9, 0.91, 0.94)
y <- c(0.37, 0.06, 0.2, 0.27, 0.57, 0.66, 0.63, 0.91, 0.9, 0.94)
tau(x,y)
# 0.7777778
tauAP(x,y)
# 0.7491182
```
In `tauAP` it is important to use the correct sorting order. By default, items are sorted in decreasing order, as should be for instance if the scores represent system effectiveness. When they should be sorted in increasing order, `decreasing` should be set to `FALSE`:

```r
# these two calls are equivalent
tauAP(x,y)
# 0.7491182
tauAP(-x,-y, decreasing = FALSE)
# 0.7491182
```

`tau_a` and `tauAP_a` are versions to use when `x` represents a true ranking without ties, and `y` represents a ranking estimated by an observer who is allowed to produce ties. They can be used as a measure of *accuracy* of the observer with respect to the true ranking.

```r
y <- round(y*5)/5 # simulate ties in y
tau_a(x,y)
# 0.7111111
tauAP_a(x,y)
# 0.6074515
```
`tau_b` and `tauAP_b` are versions to use under the assumption that both `x` and `y` represent rankings estimated by two observers who may produce ties. They can be used as a measure of *agreement* between the observers:

```r
x <- round(x*5)/5 # simulate ties in x as well
tau_b(x,y)
# 0.75
tauAP_b(x,y)
# 0.6269841
```

## License

`ircor` is released under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG
## ircor 1.0 (16-Aug-2017)

* Initial release.


tests/testthat.R
library(testthat)
library(ircor)

test_check("ircor")


tests/testthat/test_tau.R
context("tau")

test_that("tau", {
  expect_equal(tau(set1$x, set1$y), .9111, tolerance = 5e-5)
  expect_equal(tau(set2$x, set2$y), .2889, tolerance = 5e-5)
  expect_equal(tau(set3$x, set3$y), -.6889, tolerance = 5e-5)

  # check symmetry
  expect_equal(tau(set1$y, set1$x), .9111, tolerance = 5e-5)
  expect_equal(tau(set2$y, set2$x), .2889, tolerance = 5e-5)
  expect_equal(tau(set3$y, set3$x), -.6889, tolerance = 5e-5)
})

test_that("tau_a", {
  # check that it's the same as tau when there are no ties
  expect_equal(tau_a(set1$x, set1$y), .9111, tolerance = 5e-5)
  expect_equal(tau_a(set2$x, set2$y), .2889, tolerance = 5e-5)
  expect_equal(tau_a(set3$x, set3$y), -.6889, tolerance = 5e-5)
  # and symmetry
  expect_equal(tau_a(set1$y, set1$x), .9111, tolerance = 5e-5)
  expect_equal(tau_a(set2$y, set2$x), .2889, tolerance = 5e-5)
  expect_equal(tau_a(set3$y, set3$x), -.6889, tolerance = 5e-5)

  # now with ties in y
  expect_equal(tau_a(set1$x, set1$y.ties), .8889, tolerance = 5e-5)
  expect_equal(tau_a(set2$x, set2$y.ties), .3333, tolerance = 5e-5)
  expect_equal(tau_a(set3$x, set3$y.ties), -.6222, tolerance = 5e-5)
})

test_that("tau_b", {
  # check that it's the same as tau when there are no ties
  expect_equal(tau_b(set1$x, set1$y), .9111, tolerance = 5e-5)
  expect_equal(tau_b(set2$x, set2$y), .2889, tolerance = 5e-5)
  expect_equal(tau_b(set3$x, set3$y), -.6889, tolerance = 5e-5)
  # and symmetry
  expect_equal(tau_b(set1$y, set1$x), .9111, tolerance = 5e-5)
  expect_equal(tau_b(set2$y, set2$x), .2889, tolerance = 5e-5)
  expect_equal(tau_b(set3$y, set3$x), -.6889, tolerance = 5e-5)

  # now with ties
  expect_equal(tau_b(set1$x.ties, set1$y.ties), .9398, tolerance = 5e-5)
  expect_equal(tau_b(set2$x.ties, set2$y.ties), .3765, tolerance = 5e-5)
  expect_equal(tau_b(set3$x.ties, set3$y.ties), -.6510, tolerance = 5e-5)
  # and symmetry
  expect_equal(tau_b(set1$y.ties, set1$x.ties), .9398, tolerance = 5e-5)
  expect_equal(tau_b(set2$y.ties, set2$x.ties), .3765, tolerance = 5e-5)
  expect_equal(tau_b(set3$y.ties, set3$x.ties), -.6510, tolerance = 5e-5)
})


tests/testthat/test_tauAP.R
context("tauAP")

test_that("tauAP", {
  expect_equal(tauAP(set1$x, set1$y), .8519, tolerance = 5e-5)
  expect_equal(tauAP(set2$x, set2$y), .2504, tolerance = 5e-5)
  expect_equal(tauAP(set3$x, set3$y), -.6971, tolerance = 5e-5)
})

test_that("tauAP_a", {
  # check that it's the same as tauAP when there are no ties
  expect_equal(tauAP_a(set1$x, set1$y), .8519, tolerance = 5e-5)
  expect_equal(tauAP_a(set2$x, set2$y), .2504, tolerance = 5e-5)
  expect_equal(tauAP_a(set3$x, set3$y), -.6971, tolerance = 5e-5)

  # now with ties in y
  expect_equal(tauAP_a(set1$x, set1$y.ties), .7454, tolerance = 5e-5)
  expect_equal(tauAP_a(set2$x, set2$y.ties), .2692, tolerance = 5e-5)
  expect_equal(tauAP_a(set3$x, set3$y.ties), -.5558, tolerance = 5e-5)
})

test_that("tauAP_b", {
  # check that it's the same as symmetric tauAP when there are no ties
  expect_equal(tauAP_b(set1$x, set1$y), .8333, tolerance = 5e-5)
  expect_equal(tauAP_b(set2$x, set2$y), .2880, tolerance = 5e-5)
  expect_equal(tauAP_b(set3$x, set3$y), -.668, tolerance = 5e-5)

  # now with ties
  expect_equal(tauAP_b(set1$x.ties, set1$y.ties), .7321, tolerance = 5e-5)
  expect_equal(tauAP_b(set2$x.ties, set2$y.ties), .297, tolerance = 5e-5)
  expect_equal(tauAP_b(set3$x.ties, set3$y.ties), -.7047, tolerance = 5e-5)
  # and symmetry
  expect_equal(tauAP_b(set1$y.ties, set1$x.ties), .7321, tolerance = 5e-5)
  expect_equal(tauAP_b(set2$y.ties, set2$x.ties), .297, tolerance = 5e-5)
  expect_equal(tauAP_b(set3$y.ties, set3$x.ties), -.7047, tolerance = 5e-5)
})


R/tau.R
#' Kendall \eqn{\tau} Rank Correlation Coefficients
#'
#' \code{tau} is the rank correlation coefficient by Kendall, where neither vector can contain tied
#' items. \code{tau_a} and \code{tau_b} are the versions developed to cope with ties under the
#' scenarios of accuracy and agreement, respectively. See the references for details.
#'
#' @param x a numeric vector. In \code{tau_a} this is the vector of true scores.
#' @param y a numeric vector of the same length as \code{x}. In \code{tau_a} this is the vector of
#'   estimated scores.
#'
#' @return The correlation coefficient.
#' @references M.G. Kendall (1970). Rank Correlation Methods. Charles Griffin & Company
#'   Limited.
#' @seealso \code{\link{tauAP}} for AP correlation coefficients.
#' @examples
#' # No ties
#' x <- c(0.67, 0.45, 0.29, 0.12, 0.57, 0.24, 0.94, 0.75, 0.08, 0.54)
#' y <- c(0.48, 0.68, 0.32, 0.09, 0.06, 0.61, 0.87, 0.22, 0.44, 0.84)
#' tau(x, y)
#' tau_a(x,y) # same as tau
#' tau_b(x,y) # same as tau
#'
#' # Ties in y
#' y <- round(y, 1)
#' tau_a(x, y)
#' tau_b(x, y)
#'
#' # Ties in x too
#' x <- round(x, 1)
#' tau_b(x, y)
#' @export
tau <- function(x, y) {
  check(x, y)

  n <- length(x)
  numerator <- 0
  for(i in seq(1, n-1)) {
    for(j in seq(i+1, n)) {
      sx <- sign(x[i] - x[j])
      sy <- sign(y[i] - y[j])

      numerator <- numerator + sx * sy
    }
  }
  nn <- n*(n-1)/2
  numerator / nn
}

#' @rdname tau
#' @export
tau_a <- function(x, y) {
  check_a(x, y)

  n <- length(x)
  numerator <- 0
  for(i in seq(1, n-1)) {
    for(j in seq(i+1, n)) {
      sx <- sign(x[i] - x[j])
      sy <- sign(y[i] - y[j])

      numerator <- numerator + sx * sy
    }
  }
  nn <- n*(n-1)/2
  numerator / nn
}

#' @rdname tau
#' @export
tau_b <- function(x, y) {
  check_b(x, y)

  n <- length(x)
  numerator <- 0
  tx <- ty <- 0
  for(i in seq(1, n-1)) {
    for(j in seq(i+1, n)) {
      sx <- sign(x[i] - x[j])
      sy <- sign(y[i] - y[j])

      numerator <- numerator + sx * sy
      if(sx == 0)
        tx <- tx +1
      if(sy == 0)
        ty <- ty +1
    }
  }
  nn <- n*(n-1)/2
  numerator / sqrt(nn-tx) / sqrt(nn-ty)
}


R/check.R
#' Check Arguments to Correlation Functions
#'
#' These are auxiliary functions to check that two variables are of the correct type to compute
#' correlation coefficients. If they are not, the execution is \code{\link{stop}}ped.
#'
#' \itemize{
#'   \item \code{check_types}: both \code{x} and \code{y} are numeric vectors of the same length.
#'   \item \code{check} : \code{check_types} and neither \code{x} nor \code{y} contains ties.
#'   \item \code{check_a}: \code{check_types} and \code{x} does not contain ties.
#'   \item \code{check_b}: same as \code{check_types}.
#' }
#'
#' @param x the first argument.
#' @param y the second argument.
#'
#' @return Nothing, called for its side effect.
#' @noRd
check_types <- function(x, y) {
  if(!is.numeric(x) || !is.atomic(x) || !is.vector(x))
    stop("'x' must be a numeric vector", call. = FALSE)
  if(!is.numeric(y) || !is.atomic(y) || !is.vector(y))
    stop("'y' must be a numeric vector", call. = FALSE)
  if(length(x) != length(y))
    stop("'x' and 'y' must be of the same length", call. = FALSE)
}

has.ties <- function(x) {
  length(x) != length(table(x))
}

#' @rdname check_types
check <- function(x, y) {
  check_types(x,y)

  if(has.ties(x))
    stop("'x' contains ties.", call. = FALSE)
  if(has.ties(y))
    stop("'y' contains ties.", call. = FALSE)
}

#' @rdname check_types
check_a <- function(x, y) {
  check_types(x, y)
  if(has.ties(x))
    stop("'x' contains ties.", call. = FALSE)
}

#' @rdname check_types
check_b <- function(x, y) {
  check_types(x, y)
}



R/permuteTies.R
#' Permutations of Rankings With Ties
#'
#' Generates all the permutations of a vector of ranks, by permuting the ranks of all tied items.
#'
#' @param x a numeric vector.
#' @param decreasing logical. Should the sort order be increasing or decreasing (default)?
#'
#' @return A list with all the permuted rankings.
#' @noRd
permuteTies <- function(x, decreasing = TRUE) {
  if(decreasing)
    x <- -x

  ranks <- rank(x, ties.method = "first") # each group contains all integral ranks to permute
  groups <- rank(x, ties.method = "min") # each group contains only the 1st rank within the group
  doPermuteTies(ranks, groups)
}

#' Recursive Generation of Permuted Rankings
#'
#' This function computes recursively the list of all permutations of a given ranking.
#'
#' @param ranks integral ranks of pending items.
#' @param groups tied-ranks (non-integral) of the pending items.
#' @param prev ranks given to the previous items.
#' @param l list to collect all permutations.
#'
#' @return A list with all the permuted rankings.
#' @noRd
doPermuteTies <- function(ranks, groups,  prev = numeric(0), l = list()){
  if(length(ranks) == 1) {
    # only one item remaining. Concatenate and add
    l <- append(l, list( c(prev, ranks) ))
  } else {
    # work out the next item. Obtain tied items and their ranks
    sameGroup <- which(groups == groups[1])
    groupRanks <- ranks[sameGroup]
    # recursively process the rest, rotating items within the same group
    for(i in seq_along(sameGroup)) {
      ranks2 <- ranks[-1]
      ranks2[sameGroup[-1]-1] <- groupRanks[-i]
      groups2 <- groups[-1]
      prev2 <- c(prev, groupRanks[i])
      l <- doPermuteTies(ranks2, groups2, prev2, l)
    }
  }
  l
}


R/tauAP.R
#' AP Rank Correlation Coefficients
#'
#' \code{tauAP} is the AP rank correlation coefficient by Yilmaz et al., where neither vector can
#' contain tied items. \code{tauAP_a} and \code{tauAP_b} are the versions developed by Urbano and
#' Marrero to cope with ties under the scenarios of accuracy and agreement, respectively. See the
#' references for details.
#'
#' Note that the sorting order is decreasing by default, as should be for instance if the scores
#' represent the effectiveness of systems. When the sorting order is ascending, as is for instance when the vectors represent ranks, the parameter
#' \code{decreasing} must be set to \code{FALSE}.
#'
#' @param x a numeric vector. This is the vector of true scores.
#' @param y a numeric vector of the same length as \code{x}. This is the vector of estimated scores.
#' @param decreasing logical. Should the sort order be increasing or decreasing (default)?
#'
#' @return The correlation coefficient.
#' @references E. Yilmaz, J.A. Aslam and S. Robertson (2008). A New Rank Correlation Coefficient for
#'   Information Retrieval. ACM SIGIR.
#'
#'   J. Urbano and M. Marrero (2017). The Treatment of Ties in AP Correlation. ACM ICTIR.
#' @seealso \code{\link{tau}} for Kendall correlation coefficients.
#' @examples
#' # No ties
#' x <- c(0.67, 0.45, 0.29, 0.12, 0.57, 0.24, 0.94, 0.75, 0.08, 0.54)
#' y <- c(0.48, 0.68, 0.32, 0.09, 0.06, 0.61, 0.87, 0.22, 0.44, 0.84)
#' tauAP(x, y)
#' tauAP_a(x,y) # same as tauAP
#'
#' # Ties in y
#' y <- round(y, 1)
#' tauAP_a(x, y)
#' tauAP_b(x, y)
#'
#' # Ties in x too
#' x <- round(x, 1)
#' tauAP_b(x, y)
#'
#' # Set decreasing to FALSE when x and y already represent ranks
#' x <- rank(-x)
#' y <- rank(-y)
#' tauAP_b(x, y, FALSE) # same as above
#' @export
tauAP <- function(x, y, decreasing = TRUE) {
  check(x, y)

  if(decreasing)
    return(tauAP(-x, -y, decreasing = FALSE))

  n <- length(x)
  rx <- rank(x)
  ry <- rank(y)

  numerator <- 0
  for(i in seq(1, n-1)) {
    for(j in seq(i+1, n)) {
      sx <- sign(x[i] - x[j])
      sy <- sign(y[i] - y[j])

      if(sx == sy)
        # since we don't traverse in sorted order, divide by the max rank
        numerator <- numerator + 1 / ( max(ry[i], ry[j]) -1 )
    }
  }
  2* numerator / (n-1) -1
}

#' @rdname tauAP
#' @export
tauAP_a <- function(x, y, decreasing = TRUE) {
  check_a(x, y)

  if(decreasing)
    return(tauAP_a(-x, -y, decreasing = FALSE))

  n <- length(x)
  rx <- rank(x)
  ry <- rank(y, ties.method = "first")

  p <- rank(y, ties.method = "min")
  t <- sapply(p, function(pp) sum(p == pp))

  c_all <- 0
  # term I: concordants above the group
  for(i in which(p != 1)) { # ignore first item('s group)
    c_above <- 0
    for(j in which(p < p[i])) {
      sx <- sign(rx[i] - rx[j])
      sy <- sign(ry[i] - ry[j])

      if(sx == sy)
        c_above <- c_above +1
    }
    s_above <- sum( 1 / (p[i] + seq(t[i]) -2) / t[i] )

    c_all <- c_all + c_above * s_above
  }
  # term II: concordants within the group across permutations
  for(i in seq_along(p)) {
    c_within <- 0
    for(k in seq(length.out = t[i]-1))
      c_within <- c_within + k / (p[i] + k -1)
    c_within <- c_within / 2 / t[i]

    c_all <- c_all + c_within
  }
  2 / (n-1) * c_all -1
}

#' @rdname tauAP
#' @export
tauAP_b <- function(x, y, decreasing = TRUE) {
  check_b(x, y)

  if(decreasing)
    return(tauAP_b(-x, -y, decreasing = FALSE))

  (tauAP_b_ties(x,y) + tauAP_b_ties(y,x)) / 2
}

#' Asymetric tauAP_b
#' @noRd
tauAP_b_ties <- function(x, y) {
  n <- length(x)
  rx <- rank(x)
  ry <- rank(y, ties.method = "first")

  p <- rank(y, ties.method = "min")

  c_all <- 0
  not_top <- which(p != 1)
  for(i in not_top) { # ignore the first item('s group)
    # count concordants above the pivot's tie group
    c_above <- 0
    for(j in which(p < p[i])) {
      sx <- sign(rx[i] - rx[j])
      sy <- sign(ry[i] - ry[j])

      if(sx == sy)
        c_above <- c_above +1
    }
    c_all <- c_all + c_above / (p[i]-1) # divide by p-1 instead of i-1
  }
  2 / length(not_top) * c_all -1
}
